pageserver/src/tenant/timeline.rs

```diff
@@ -3512,16 +3513,25 @@ impl Timeline {
         // min-heap (reserve space for one more element added before eviction)
         let mut heap: BinaryHeap<Hole> = BinaryHeap::with_capacity(max_holes + 1);
         let mut prev: Option<Key> = None;
-        for (next_key, _next_lsn, _size) in itertools::process_results(
-            deltas_to_compact.iter().map(|l| -> Result<_> {
-                Ok(l.clone()
-                    .downcast_delta_layer()
-                    .expect("delta layer")
-                    .load_keys(ctx)?
-                    .into_iter())
-            }),
-            |iter_iter| iter_iter.kmerge_by(|a, b| a.0 < b.0),
-        )? {
+
+        let mut all_keys = Vec::new();
+        for l in deltas_to_compact.iter() {
+            // TODO: replace this with an await once we fully go async
+            all_keys.extend(
+                Handle::current().block_on(
+                    l.clone()
+                        .downcast_delta_layer()
+                        .expect("delta layer")
+                        .load_keys(ctx),
+                )?,
+            );
```

Comment:
I wonder do the delta layers have a number of nodes we could use to pre-allocate the vec?

At worst case here it's doing `10 x keys(256MB delta file)` which would be easy to avoid, if such count would be available.

Reply:
There is no such number stored for on-disk btrees. One would have to do an entire traversal via visit, involving lots of I/O calls.

Vecs have amortized O(1) time, amortized means that each item only gets inserted roughly twice *on average*. In short, you can visualize this via the last copy copying half, the one before copying a quarter, etc, until you reach the start size of the vec. In other words, it's `1/2 + 1/4 + 1/8 + ... ≈ 1` many copies in total (multiplied by the number of items). even though the starting items get copied `log_2(n)` many times, half of the items we don't copy at all, and a quarter only gets copied once, etc.

For `extend` calls, you can model them as repeated insertions, but with the additional optimization that some doublings might be skipped entirely if e.g. the size of the allocation gets increased from 128 to 1024 via the addition of a 600 item slice.

Reply:
Yep, oki, thanks for checking. Maybe it should be added later.